### PR TITLE
feat(KSA): DoNotVersion for Initialize Mutation

### PR DESCRIPTION
**Pull request by texastony**
_Issue #, if available:_

_Description of changes:_
1. Implement `doNotVersion` behavior in Initialize Mutation
2. Refactor Initialize, Apply, and Mutations.dfy to simplify Pre/Post conditions for Mutate

In a separate PR,
I will revise the Documentation trait for `doNotVersion` to explain
when it is a good idea to Version and when it is not needed.

_Squash/merge commit message, if applicable:_

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/ApplyMutation.dfy [hunk 1]
```diff
@@ -389,19 +381,20 @@ module {:options "/functionSyntax:4" } InternalApplyMutation {
...
+          //   mutationToApply.Terminal.customEncryptionContext
+          // );
+
+          // var mutatedItem :- Mutations.ReEncryptHierarchicalKey(
+          //   item := item,
+          //   originalKmsArn := mutationToApply.Original.kmsArn,
+          //   terminalKmsArn := mutationToApply.Terminal.kmsArn,
+          //   terminalEncryptionContext := terminalEncryptionContext,
+          //   keyManagerStrategy := keyManagerStrategy
+          // );
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/ApplyMutation.dfy:397]:
```suggestion
```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy [hunk 2]
```diff
@@ -330,68 +323,38 @@ module {:options "/functionSyntax:4" } InternalInitializeMutation {
...
-    // --= Generate New Decrypt Only Branch Key with terminal properties
-    var maybeNewVersion := UUID.GenerateUUID();
-    var newVersion :- maybeNewVersion
-    .MapFailure(e => Types.KeyStoreAdminException(
-                    message := "Could not generate UUID for new Decrypt Only. " + e));
+    // Mutations-doNotVersion
+    // I think the cleanest way to fry this fish
+    // is to break out a new method.
+    // It takes in the Input, the read ActiveItem, and MutationsToApply
+    // It returns WriteActive & WriteVersion
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy:330]:
```suggestion
```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy [hunk 3]
```diff
@@ -330,68 +323,38 @@ module {:options "/functionSyntax:4" } InternalInitializeMutation {
...
+    // assert readItems.BeaconItem.KmsArn == MutationToApply.Original.kmsArn;
+
+    // var newBeaconKey :- Mutations.ReEncryptHierarchicalKey(
+    //   item := readItems.BeaconItem,
+    //   originalKmsArn := MutationToApply.Original.kmsArn,
+    //   terminalKmsArn := MutationToApply.Terminal.kmsArn,
+    //   terminalEncryptionContext := BeaconEncryptionContext,
+    //   keyManagerStrategy := input.keyManagerStrategy,
+    //   localOperation := "InitializeMutation"
+    // );
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy:357]:
```suggestion
```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy [hunk 4]
```diff
@@ -610,4 +568,223 @@ module {:options "/functionSyntax:4" } InternalInitializeMutation {
...
+    requires localInput.input.DoNotVersion
+    ensures
+      output.Success?
+      ==>
+        && output.value.writeVersion.mutate?
+        && |output.value.logStatements| == 2
+  {
+    // output := Failure(Types.UnsupportedFeatureException(message := "WIP"));
+    // return output;
+
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/InitializeMutation.dfy:733]:
```suggestion
 ```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy [hunk 5]
```diff
@@ -218,49 +219,85 @@ module {:options "/functionSyntax:4" } Mutations {
...
     nameonly terminalEncryptionContext: Structure.BranchKeyContext,
-    nameonly keyManagerStrategy: KmsUtils.keyManagerStrat,
-    nameonly localOperation: string := "ApplyMutation"
+    nameonly keyManagerStrategy: KmsUtils.keyManagerStrat
+  )
+  {
+    ghost predicate Pre()
+    {
+      && Structure.EncryptedHierarchicalKey?(item)
+         // && KMS.IsValid_KeyIdType(terminalKmsArn) // Should be met by ValidKmsArn?(terminalKmsArn)
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy:234]:
```suggestion
```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy [hunk 6]
```diff
@@ -218,49 +219,85 @@ module {:options "/functionSyntax:4" } Mutations {
...
+    }
+  }
+
+  method {:isolate_assertions} ReEncryptHierarchicalKey(
+    nameonly input: ReEncryptHierarchicalKeyInput,
+    // nameonly item: Types.AwsCryptographyKeyStoreTypes.EncryptedHierarchicalKey,
+    // nameonly originalKmsArn: string,
+    // nameonly terminalKmsArn: string,
+    // nameonly terminalEncryptionContext: Structure.BranchKeyContext,
+    // nameonly keyManagerStrategy: KmsUtils.keyManagerStrat,
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy:258]:
```suggestion
```

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy [hunk 7]
```diff
@@ -218,49 +219,85 @@ module {:options "/functionSyntax:4" } Mutations {
...
+    // // requires KMS.IsValid_KeyIdType(terminalKmsArn) // Should be met by ValidKmsArn?(terminalKmsArn)
+    // requires KMSKeystoreOperations.AttemptReEncrypt?(item.EncryptionContext, terminalEncryptionContext)
+    // requires KmsArn.ValidKmsArn?(originalKmsArn) && KmsArn.ValidKmsArn?(terminalKmsArn)
+    // requires item.KmsArn == originalKmsArn
+    // requires keyManagerStrategy.ValidState()
+    // modifies
+    //   match keyManagerStrategy
+    //   case reEncrypt(km) => km.kmsClient.Modifies
+    //   case decryptEncrypt(kmD, kmE) => kmD.kmsClient.Modifies + kmE.kmsClient.Modifies
+    // ensures keyManagerStrategy.ValidState()
```
Comment (texastony) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/Mutations.dfy:276]:
```suggestion
 ```

**Review by texastony — COMMENTED** [review]
Remove comments

**Review by josecorella — APPROVED** [review]

**Pull request merged**